File: spambot/SPAMFramework/Packet/GetActionPacket.cpp
#include "GetActionPacket.h"

GetActionPacket::GetActionPacket(int pot_size, int num_board_cards,
		std::vector<Card*> board_cards, int num_last_actions,
		std::vector<std::string> last_actions, int num_legal_actions,
		std::vector<std::string> legal_actions, double time_bank) {
	m_pot_size = pot_size;
	m_num_board_cards = num_board_cards;
	m_board_cards = board_cards;
	m_num_last_actions = num_last_actions;
	m_last_actions = last_actions;
	m_num_legal_actions = num_legal_actions;
	m_legal_actions = legal_actions;
	m_time_bank = time_bank;
}

int GetActionPacket::NumBoardCardsToStreet(int num_cards) {
	switch (num_cards) {
	case 0:
		return GameState::PREFLOP;
	case 3:
		return GameState::FLOP;
	case 4:
		return GameState::TURN;
	case 5:
		return GameState::RIVER;
	}

	return GameState::GAMEOVER;
}

void GetActionPacket::UpdateGameState(GameState *gs) {
	gs->SetPotSize(m_pot_size);
	gs->SetCurrentStreet(NumBoardCardsToStreet(m_num_board_cards));
	gs->SetBoardCards(m_board_cards);

	gs->GetSelfHistory()->SetLastActions(m_last_actions);
	gs->GetVillainHistory()->SetLastActions(m_last_actions);

	gs->GetSelfHistory()->LookForDiscards();
	
	// When streets change
	if (gs->GetSelfHistory()->GetLastStreet() != gs->GetCurrentStreet()) {
		gs->GetSelfHistory()->ResetForNewStreet(gs->GetCurrentStreet());
	}
	if (gs->GetVillainHistory()->GetLastStreet() != gs->GetCurrentStreet()) {
		gs->GetVillainHistory()->ResetForNewStreet(gs->GetCurrentStreet());
	}

	// Thought that this should go before street change logic so that
	// these updates can be reset to 0 for each street
	gs->GetSelfHistory()->UpdateBetAmounts();
	gs->GetVillainHistory()->UpdateBetAmounts();

	gs->SetLegalActions(m_legal_actions);
	
	gs->SetTimeLeft(m_time_bank);
}

File: java_bot/src/pokerbots/spam_framework/packet/GetActionPacket.java
package pokerbots.spam_framework.packet;

import java.util.ArrayList;

import pokerbots.spam_framework.Card;
import pokerbots.spam_framework.GameState;

public class GetActionPacket extends Packet {

	private int potSize;
	private int numBoardCards; 
	private ArrayList<Card> boardCards;
	private int numLastActions; 
	private ArrayList<String> lastActions;
	private int numLegalActions;
	private ArrayList<String> legalActions;
	private double timeBank;
	
	public GetActionPacket(int potSize, int numBoardCards, 
			ArrayList<Card> boardCards, int numLastActions, 
			ArrayList<String> lastActions, int numLegalActions,
			ArrayList<String> legalActions, double timeBank) {
		
		this.potSize = potSize;
		this.numBoardCards = numBoardCards;
		this.boardCards = boardCards;
		this.numLastActions = numLastActions;
		this.lastActions = lastActions;
		this.numLegalActions = numLegalActions;
		this.legalActions = legalActions;
		this.timeBank = timeBank;
	}
	
	private static int numBoardCardsToStreet(int numCards) {
		switch(numCards){
		case 0:
			return GameState.PREFLOP;
		case 3:
			return GameState.FLOP;
		case 4:
			return GameState.TURN;
		case 5:
			return GameState.RIVER;
		}
		
		return GameState.GAMEOVER;
	}
	
	@Override
	public void updateGameState(GameState gameState) {
		gameState.setPotSize(potSize);
		gameState.setCurrentStreet(GetActionPacket
				.numBoardCardsToStreet(numBoardCards));
		gameState.setBoardCards(boardCards);
		
		// TODO: Need to separate the two action histories
		gameState.getSelfHistory().setLastActions(lastActions);
		gameState.getVillainHistory().setLastActions(lastActions);
		
		gameState.setLegalActions(legalActions);
		
		gameState.setTimeLeft(timeBank);
	}

}
